kratos/python_scripts/restart_utility.py

```diff
@@ -170,6 +191,46 @@ def CreateOutputFolder(self):
                 os.makedirs(folder_path)
             self.model_part.GetCommunicator().GetDataCommunicator().Barrier()
 
+    def GetRestartFiles(self):
+        """Return a dictionary of stepID - restart_file_list dictionary that stores sets of restart files for each step."""
+        restart_path    = Path( self.__GetFolderPathSave() )
+        restart_files   = {}
+        if restart_path.is_dir():
+            for entry in restart_path.glob("**/*"):
+                if entry.is_file() and self.__IsRestartFile( entry.name ):
+                    label = self.__ExtractFileLabel(entry.name) # Get thread ID and step ID
+                    if label[1] in restart_files:               # Check if this step has entries already
+                        restart_files[label[1]].append( entry.name )
+                    else:
+                        restart_files[label[1]] = [ entry.name ]
+
+        return restart_files
+
+    def ClearObsoleteRestartFiles(self):
+        """Delete restart files that are no longer needed."""
+        if self.number_of_restart_files > -1:               # <-- number of restart files is limited
+            number_of_obsolete_files = len(self.restart_files) - self.number_of_restart_files
+            for _ in range(number_of_obsolete_files):
+
+                # Get oldest restart file set
+                oldest_step_id, oldest_file_set = min( self.restart_files.items(), key=lambda item: float(item[0]) )
+
+                # Try to delete every file in the set
+                for file_name in oldest_file_set:
+                    file_path = os.path.join( self.__GetFolderPathSave(), file_name )
+                    try:
+                        if os.path.isfile( file_path ):
+                            os.remove( file_path )
+                    except OSError:
+                        message =   'Failed to delete restart file "'
+                        message +=  file_path
+                        message += '"'
+                        print( message )                # <-- TODO: decide whether to throw a non-blocking exception or display a warning
+                        #raise Exception(message)       #
```

Comment:
here you can use [this](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/kratos_utilities.py#L9) instead

this has proven to work fine for us, both in serial and in MPI